### PR TITLE
[SPARK-26003] Improve SQLAppStatusListener.aggregateMetrics performance

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

In `SQLAppStatusListener.aggregateMetrics`, we use the `metricIds` only to filter the relevant metrics. And this is a Seq which is also sorted. When there are many metrics involved, this can be pretty inefficient. The PR proposes to use a Set for it.

## How was this patch tested?

NA

### CODE REVIEW

**Comment by mgaido91** [comment]
cc @cloud-fan @vanzin

**Comment by cloud-fan** [comment]
LGTM, also cc @gengliangwang

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 1]
```diff
@@ -159,7 +159,7 @@ class SQLAppStatusListener(
   }
 
   private def aggregateMetrics(exec: LiveExecutionData): Map[Long, String] = {
-    val metricIds = exec.metrics.map(_.accumulatorId).sorted
+    val metricIds = exec.metrics.map(_.accumulatorId).toSet
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: Actually this one can be merged into `metricTypes`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala [hunk 2]
```diff
@@ -159,7 +159,7 @@ class SQLAppStatusListener(
...
     val metricTypes = exec.metrics.map { m => (m.accumulatorId, m.metricType) }.toMap
     val metrics = exec.stages.toSeq
       .flatMap { stageId => Option(stageMetrics.get(stageId)) }
```
Comment (gengliangwang) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: Consider also change the following `flatMap` / `filter`  / `groupBy` into `while` loop

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: not sure what you mean here. Why should we use a `while` loop?

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: If the metrics is large, then using a while loop can reduce the number of traversal loops. And it is not complicated to do it in the code here.

Comment (gengliangwang) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: I am also fine with the current code here.

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/ui/SQLAppStatusListener.scala]: yes, we can save 1 traversal, but I am not sure it is worth honestly... This approach seems cleaner to me.

**Review by gengliangwang — APPROVED** [review]
LGTM

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**